Tests/MongoSwiftTests/CodecTests.swift

```diff
@@ -714,41 +718,51 @@ final class CodecTests: MongoSwiftTestCase {
         expect(try decoder.decode(AnyBSONStruct.self,
                                   from: wrappedMinKey.canonicalExtendedJSON).x.value as? MinKey).to(equal(minKey))
+
+        // NSNull
+        expect(
+            try decoder.decode(AnyBSONStruct.self, from: ["x": NSNull()]).x
+        ).to(equal(AnyBSONValue(NSNull())))
+
+        expect(try encoder.encode(AnyBSONStruct(NSNull()))).to(equal(["x": NSNull()]))
     }
 
-    struct OptionalAnyBSONWrapper: Codable {
-        let val: AnyBSONValue?
+    fileprivate struct IncorrectTopLevelEncode: Encodable {
+        let x: AnyBSONValue
+
+        // An empty encode here is incorrect.
+        func encode(to encoder: Encoder) throws {}
 
-        init(_ value: BSONValue?) {
-            self.val = AnyBSONValue(ifPresent: value)
+        init(_ x: BSONValue) {
+            self.x = AnyBSONValue(x)
         }
     }
 
-    func testOptionalAnyBSONValue() throws {
-        let encoder = BSONEncoder()
-        let decoder = BSONDecoder()
+    fileprivate struct CorrectTopLevelEncode: Encodable {
+        let x: IncorrectTopLevelEncode
 
-        // non-nil values
-        expect(try encoder.encode(OptionalAnyBSONWrapper(5))).to(equal(["val": 5]))
+        // swiftlint:disable nesting
+        enum CodingKeys: CodingKey {
+            case x
+        }
 
-        let doc1: Document = ["y": 1]
-        expect(try encoder.encode(OptionalAnyBSONWrapper(doc1))).to(equal(["val": doc1]))
-        let doc2: Document = ["x": 1, "y": nil]
-        expect(try encoder.encode(OptionalAnyBSONWrapper(doc2))).to(equal(["val": doc2]))
-        let arr1: [BSONValue] = [1, 2, "hi"]
-        expect(try encoder.encode(OptionalAnyBSONWrapper(arr1))).to(equal(["val": arr1]))
+        // An empty encode here is incorrect.
+        func encode(to encoder: Encoder) throws {
+            var container = encoder.container(keyedBy: CodingKeys.self)
+            try container.encode(x, forKey: .x)
+        }
 
-        // an array with a nil
-        let arr2: [BSONValue?] = [1, "hi", nil]
-        expect(try encoder.encode(OptionalAnyBSONWrapper(arr2))).to(equal(["val": arr2]))
+        init(_ x: BSONValue) {
+            self.x = IncorrectTopLevelEncode(x)
+        }
+    }
 
-        // an array with a non-BSONValue
-        let arr3: [Any?] = [1, "hi", nil, Int16(4)]
-        expect(try arr3.encode(to: DocumentStorage(), forKey: "arr3")).to(throwError())
+    func testIncorrectEncodeFunction() {
+        let encoder = BSONEncoder()
 
-        // nil value
-        expect(try encoder.encode(OptionalAnyBSONWrapper(nil))).to(beNil())
+        // A top-level `encode()` problem should throw an error, but any such issues deeper in the recursion should not.
+        expect(try encoder.encode(IncorrectTopLevelEncode(NSNull()))).to(throwError())
+        expect(try encoder.encode(CorrectTopLevelEncode(NSNull()))).toNot(throwError())
```

Comment:
can we also compare this result against what we expect it to be?

Reply:
perhaps also state that this is the behavior of `JSONEncoder`

Reply:
Done.

Reply:
nit: this is redundant with the next check now, since that will also fail if encoding throws
```suggestion
```